### PR TITLE
Port to 1.21

### PR DESCRIPTION
**Pull request by unilock**
General TODO:
- [x] Fix FabRefl stuff (`ap` subproject, `jar` task, etc.)
- [x] Port FScript to 1.21
- [ ] Test EVERYTHING
- [ ] ?????
- [ ] Forgery :/

More Specific TODO:
  - [x] Fix a few mixin injection failures (these should all be valid!?)
    - [x] `com.unascribed.fabrication.mixin.b_utility.legacy_command_syntax.MixinEntitySummonArgumentType`
      - [x] `net/minecraft/command/argument/RegistryEntryArgumentType;parseAsNbt(Lcom/mojang/brigadier/StringReader;)Lnet/minecraft/nbt/NbtElement;` > `Lnet/minecraft/util/Identifier;fromCommandInput(Lcom/mojang/brigadier/StringReader;)Lnet/minecraft/util/Identifier;`
    - [x] `com.unascribed.fabrication.mixin.e_mechanics.swap_conflicting_enchants.MixinAnvilScreenHandler`
      - [x] `net/minecraft/screen/AnvilScreenHandler;updateResult()V` > `Lnet/minecraft/enchantment/Enchantment;canBeCombined(Lnet/minecraft/registry/entry/RegistryEntry;Lnet/minecraft/registry/entry/RegistryEntry;)Z`
      - [x] `net/minecraft/screen/AnvilScreenHandler;updateResult()V` > `Lnet/minecraft/enchantment/EnchantmentHelper;getEnchantments(Lnet/minecraft/item/ItemStack;)Lnet/minecraft/component/type/ItemEnchantmentsComponent;`      
  - [ ] Properly undo Enchantment modifications (apparently modifications to Enchantments persist across datapack reloads, even when the changes are not re-applied post-reload - maybe we need to "invalidate" certain dynamic registry entries...?)
  - [ ] Possibly remove dependency on Fabric Networking API

### CODE REVIEW

**Comment by ArsiTheFox** [comment]
Hopefully: https://github.com/FalsehoodMC/Fabrication/issues/688 is fixed

**Comment by SFort** [comment]
FScript 3.2.4 should be fine it's on the maven repo, very untested as i'm outa time for rn

**Comment by unilock** [comment]
@ArsiTheFox
> Hopefully: https://github.com/FalsehoodMC/Fabrication/issues/688 is fixed

Yes, I can confirm `grindstone_disenchanting` now works as intended with this PR.

@SFort
> FScript 3.2.4 should be fine it's on the maven repo, very untested as i'm outa time for rn

`runClient` does run the game successfully with the FScript dependency enabled now, at least :p

**Comment by unilock** [comment]
If anyone wants to help thoroughly test *everything,* dev builds are now available via GitHub Actions: https://github.com/unilock/Fabrication/actions?query=branch%3A3.0%2F1.21

**Comment by SFort** [comment]
> @ArsiTheFox
> 
> > Hopefully: #688 is fixed
> 
> Yes, I can confirm `grindstone_disenchanting` now works as intended with this PR.


are you sure? it's forgery. fabrication pretty much has no open defects/conflicts because i actually fix those.
if you did fix it on forgery pretty please pr a backport ;p

**Comment by SFort** [comment]
that is actually how a suprising amount of issues got fixed, a port demanded a recode and the recode got backported. it's actually why 1.17 forgery exists, it just started working at some point. (granted there was a fair bit of effort put in prior and afterwords, but dumb luck is what got it to boot)

**Comment by unilock** [comment]
> are you sure? it's forgery.

Ah, well, Forgery currently doesn't exist at all for 1.21, so I can't say for sure.  
I guess I should poke at that next, since I'm all but stumped on the current remaining issues...

File: .github/workflows/build.yml
Comment (SFort) [inline .github/workflows/build.yml:1]: just making a note to myself to remove this in merge

File: build.gradle [hunk 1]
```diff
@@ -77,26 +76,27 @@ dependencies {
 	modImplementation "net.fabricmc:fabric-loader:${project.loader_version}"
 
 	modImplementation fabricApi.module('fabric-api-base', project.fabric_version)
+    modImplementation fabricApi.module('fabric-networking-api-v1', project.fabric_version)
```
Comment (SFort) [inline build.gradle]: remove either pre/post merge

File: src/main/java/com/unascribed/fabrication/loaders/LoaderItemDespawn.java [hunk 2]
```diff
@@ -52,19 +52,19 @@ public void load(Path configDir, QDIni config, boolean loadError) {
 		for (String k : config.keySet()) {
 			ParsedTime time = ParsedTime.getFrom(config, k);
 			if (time instanceof ParsedTime.Unset) continue;
-			if (k.startsWith("@enchantments.")) {
+			/*if (k.startsWith("@enchantments.")) {
```
Comment (SFort) [inline src/main/java/com/unascribed/fabrication/loaders/LoaderItemDespawn.java:55]: TODO?

File: src/main/java/com/unascribed/fabrication/logic/CanHitUtil.java [hunk 3]
```diff
@@ -1,63 +1,59 @@
...
+				try {
+					UUID id = UUID.fromString(s);
+					if (entity.getUuid().equals(id)) {
 						return true;
 					}
+					continue;
+				} catch (IllegalArgumentException ex) {}
+			}
+			if (s.startsWith("@")) {
+				// TODO: EntitySelector.basePredicate no longer exists
```
Comment (SFort) [inline src/main/java/com/unascribed/fabrication/logic/CanHitUtil.java:32]: another TODO

File: src/main/java/com/unascribed/fabrication/mixin/_general/sync/MixinClientPlayNetworkHandler.java
Comment (SFort) [inline src/main/java/com/unascribed/fabrication/mixin/_general/sync/MixinClientPlayNetworkHandler.java:1]: i need to look at this later
alongside
MixinServerCommonNetworkHandler.java
MixinCustomPayloadC2SPacket.java and  MixinCustomPayloadS2CPacket.java

File: src/main/java/com/unascribed/fabrication/mixin/a_fixes/fix_charm_amethyst_dupe/MixinClearItemFrames.java [hunk 4]
```diff
@@ -1,47 +1,37 @@
...
-	}
-
-	@FabInject(at=@At("HEAD"), method="handleAttackEntity(Lnet/minecraft/class_1657;Lnet/minecraft/class_1937;Lnet/minecraft/class_1268;Lnet/minecraft/class_1297;Lnet/minecraft/class_3966;)Lnet/minecraft/class_1269;",
-			cancellable=true)
-	public void handleAttackEntity(PlayerEntity player, World world, Hand hand, Entity entity, EntityHitResult hitResult, CallbackInfoReturnable<ActionResult> ci) {
-		if (!((SetInvisibleByCharm)entity).fabrication$isInvisibleByCharm()) {
-			ci.setReturnValue(ActionResult.PASS);
-		}
-	}
+	// TODO: Check if this is still an issue
```
Comment (SFort) [inline src/main/java/com/unascribed/fabrication/mixin/a_fixes/fix_charm_amethyst_dupe/MixinClearItemFrames.java]: Knowing how the issue was originally handled, (that being pretending it's not their problem)
i'm certain they did absolutely nothing to resolve this.

File: src/main/java/com/unascribed/fabrication/mixin/c_tweaks/fullres_banner_shields/MixinBannerBlockEntityRenderer.java [hunk 5]
```diff
@@ -73,10 +77,11 @@ private static void renderCanvasHead(MatrixStack matrices, VertexConsumerProvide
 				float w = 0.625f;
 				float h = 1.25f;
 				float z = -0.125f;
-				vc.vertex(mdl, x, y, z).color(col[0], col[1], col[2], 1).texture(minU, minV).overlay(overlay).light(light).normal(nrm, 0, 0, -1).next();
-				vc.vertex(mdl, x+w, y, z).color(col[0], col[1], col[2], 1).texture(maxU, minV).overlay(overlay).light(light).normal(nrm, 0, 0, -1).next();
-				vc.vertex(mdl, x+w, y+h, z).color(col[0], col[1], col[2], 1).texture(maxU, maxV).overlay(overlay).light(light).normal(nrm, 0, 0, -1).next();
-				vc.vertex(mdl, x, y+h, z).color(col[0], col[1], col[2], 1).texture(minU, maxV).overlay(overlay).light(light).normal(nrm, 0, 0, -1).next();
+				// TODO? colors
```
Comment (SFort) [inline src/main/java/com/unascribed/fabrication/mixin/c_tweaks/fullres_banner_shields/MixinBannerBlockEntityRenderer.java:83]: todo?

File: src/main/java/com/unascribed/fabrication/mixin/c_tweaks/legible_signs/MixinSignBlockEntityRenderer.java [hunk 6]
```diff
@@ -35,8 +36,9 @@ private static void modifySignTextColor(SignText sign, CallbackInfoReturnable<In
 					res = dc.getSignColor();
 					break;
 				default: {
-					float[] bgr = dc.getColorComponents();
-					res = Math.round(bgr[0]*255.0F) << 16 | Math.round(bgr[1]*255.0F) << 8 | Math.round(bgr[2]*255);
+					// TODO?
```
Comment (SFort) [inline src/main/java/com/unascribed/fabrication/mixin/c_tweaks/legible_signs/MixinSignBlockEntityRenderer.java:42]: todo?

File: src/main/java/com/unascribed/fabrication/mixin/d_minor_mechanics/channeling_two/MixinTridentEntity.java [hunk 7]
```diff
@@ -3,28 +3,27 @@
...
+		super(entityType, world);
 	}
 
-
-	@ModifyReturn(method="onEntityHit(Lnet/minecraft/util/hit/EntityHitResult;)V", target="Lnet/minecraft/world/World;isThundering()Z")
-	public boolean fabrication$channellingTwo(boolean bool) {
-		if (!(FabConf.isEnabled("*.channeling_two") && EnchantmentHelper.getLevel(Enchantments.CHANNELING, this.getItemStack()) > 1)) return bool;
-		return this.getWorld().hasRain(this.getBlockPos());
-	}
+	// TODO
```
Comment (SFort) [inline src/main/java/com/unascribed/fabrication/mixin/d_minor_mechanics/channeling_two/MixinTridentEntity.java:22]: TODO

File: src/main/java/com/unascribed/fabrication/mixin/f_balance/pickup_skeleton_arrows/MixinArrowEntity.java
Comment (SFort) [inline src/main/java/com/unascribed/fabrication/mixin/f_balance/pickup_skeleton_arrows/MixinArrowEntity.java:1]: was this re-created somewhere or just deleted?

File: src/main/java/com/unascribed/fabrication/mixin/i_woina/block_logo/MixinTitleScreen.java [hunk 8]
```diff
@@ -31,15 +31,15 @@ protected MixinTitleScreen(Text title) {
...
 
-	@Shadow @Final
+	@Shadow
 	private boolean doBackgroundFade;
 	@Shadow
 	private long backgroundFadeStart;
 
 	@Hijack(method="render(Lnet/minecraft/client/gui/DrawContext;IIF)V", target="Lnet/minecraft/client/gui/LogoDrawer;draw(Lnet/minecraft/client/gui/DrawContext;IF)V")
-	public boolean fabrication$drawBlockLogo() {
+	public boolean fabrication$drawBlockLogo(LogoDrawer logo, DrawContext context, int i, float f) {
```
Comment (SFort) [inline src/main/java/com/unascribed/fabrication/mixin/i_woina/block_logo/MixinTitleScreen.java]: doesn't need `int i, float f`. since Hijack is part of FabInject

File: src/main/java/com/unascribed/fabrication/mixin/i_woina/dirt_screen/MixinTitleScreen.java [hunk 9]
```diff
@@ -18,14 +21,21 @@
...
 		super(title);
 	}
 
-	@FabInject(method="render(Lnet/minecraft/client/gui/DrawContext;IIF)V", at=@At(value="INVOKE", target="Lnet/minecraft/client/gui/RotatingCubeMapRenderer;render(FF)V", shift=At.Shift.AFTER))
-	public void drawDirt(DrawContext matrices, int mouseX, int mouseY, float delta, CallbackInfo ci) {
+	@FabInject(at=@At("HEAD"), method="renderPanoramaBackground(Lnet/minecraft/client/gui/DrawContext;F)V", cancellable=true)
+	public void drawDirt(DrawContext context, float delta, CallbackInfo ci) {
 		if (FabConf.isEnabled("*.dirt_screen")) {
-			renderBackgroundTexture(matrices);
+			// TODO: check if we need to mess with atlases
```
Comment (SFort) [inline src/main/java/com/unascribed/fabrication/mixin/i_woina/dirt_screen/MixinTitleScreen.java:34]: TODO

File: src/main/java/com/unascribed/fabrication/mixin/z_combined/old_armor/MixinLivingEntity.java [hunk 10]
```diff
@@ -34,25 +20,25 @@ public MixinLivingEntity(EntityType<?> type, World world) {
...
-											entry.getValue().getId(),
-											FabRefl.EntityAttributeModifier_name(entry.getValue()),
-											(old ? ArmorMaterials.DIAMOND.getProtection(((ArmorItem)stack.getItem()).getType()) : entry.getValue().getValue())
-													* (scale ? ((stack.getMaxDamage() - stack.getDamage()) / (double) stack.getMaxDamage()) : 1),
-											EntityAttributeModifier.Operation.ADDITION))
-							: new AbstractMap.SimpleEntry<>(entry.getKey(), entry.getValue()))
-				).collect(Collector.of(ArrayListMultimap::create, (m, entry) ->m.put(entry.getKey(), entry.getValue()), (m1, m2) -> {m1.putAll(m2); return m1;}));
-
-	}
+	//TODO
```
Comment (SFort) [inline src/main/java/com/unascribed/fabrication/mixin/z_combined/old_armor/MixinLivingEntity.java:26]: TODO

**Review by SFort — CHANGES_REQUESTED** [review]
I had a bit of time so thought i'd try to  get ahead of having to review this massive pr.
made some notes to myself and pointed out some TODOs so their not forgotten.
in general it's looking to be very solid, thanks!

p.s. FabRefl exists because it's older then Shadow/Accessors. continues to exist because it's it supports fail soft. some might be thinking that i'm being pedantinc and a field should always be safe to Shadow. i wish.
Don't worry about this though, it's so rare i don't remember the last time it happened.

EDIT: also forgery is a headache i wouldn't worry about it for now, got plenty of issues as is

**Comment by SFort** [comment]
> Fix a few mixin injection failures (these should all be valid!?) 

I have looked into this a bit.

Fabrication uses Mixin's Verifier to make sure it's injections didn't do something horribly wrong. ([FabInjector#328](https://github.com/FalsehoodMC/Fabrication/blob/3.0/1.18/src/main/java/com/unascribed/fabrication/support/injection/FabInjector.java#L328))

however FabInjector in fact didn't do anything horribly wrong.
it's MixinExtras that's not playing nice with Mixin's own Verifier. 

Remove the `@Local` usage you added from `disable_prior_work_penalty` and Mixins verifier will stop crashing.

**Comment by unilock** [comment]
@SFort
> Remove the `@Local` usage you added from `disable_prior_work_penalty` and Mixins verifier will stop crashing.

That did fix the injection failures in `MixinAnvilScreenHandler` for `swap_conflicting_enchants`, but `MixinEntitySummonArgumentType` for `legacy_command_syntax` is still causing injection failures. There aren't even any mixins that target the same class for that one...

File: src/main/java/com/unascribed/fabrication/mixin/i_woina/block_logo/MixinTitleScreen.java (hunk 8 above)
Comment (unilock) [reply src/main/java/com/unascribed/fabrication/mixin/i_woina/block_logo/MixinTitleScreen.java]: This feature still doesn't work anyway :/

File: src/main/java/com/unascribed/fabrication/mixin/f_balance/pickup_skeleton_arrows/MixinAbstractSkeletonEntity.java [hunk 11]
```diff
@@ -15,8 +16,8 @@
 @Mixin(AbstractSkeletonEntity.class)
 @EligibleIf(configAvailable="*.pickup_skeleton_arrows")
 public abstract class MixinAbstractSkeletonEntity {
-	@FabInject(at=@At("RETURN"), method="createArrowProjectile(Lnet/minecraft/item/ItemStack;F)Lnet/minecraft/entity/projectile/PersistentProjectileEntity;", cancellable=true)
-	public void createArrowProjectile(ItemStack arrow, float damageModifier, CallbackInfoReturnable<PersistentProjectileEntity> cir) {
+	@FabInject(at=@At("RETURN"), method="createArrowProjectile(Lnet/minecraft/item/ItemStack;FLnet/minecraft/item/ItemStack;)Lnet/minecraft/entity/projectile/PersistentProjectileEntity;", cancellable=true)
+	public void createArrowProjectile(ItemStack arrow, float damageModifier, @Nullable ItemStack shotFrom, CallbackInfoReturnable<PersistentProjectileEntity> cir) {
```
[the remark this answers is not in the document]
Comment (unilock) [reply src/main/java/com/unascribed/fabrication/mixin/f_balance/pickup_skeleton_arrows/MixinAbstractSkeletonEntity.java]: Does this apply to `@Shadow` methods as well?

File: src/main/java/com/unascribed/fabrication/mixin/f_balance/pickup_skeleton_arrows/MixinArrowEntity.java
Comment (unilock) [reply src/main/java/com/unascribed/fabrication/mixin/f_balance/pickup_skeleton_arrows/MixinArrowEntity.java:1]: IIRC it's no longer needed; potion data *should* now be correctly set (as a component) in the ItemStack of ArrowEntities shot by Strays in vanilla Minecraft.

File: .github/workflows/build.yml
Comment (unilock) [reply .github/workflows/build.yml:1]: Why not have CI?

Comment (SFort) [reply .github/workflows/build.yml:1]: i don't have a particular grudge against CI, just don't see much use to keeping it.
do keep in on the branch while the PR is ongoing though, it's usefull for this stage

File: src/main/java/com/unascribed/fabrication/mixin/f_balance/pickup_skeleton_arrows/MixinAbstractSkeletonEntity.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (SFort) [reply src/main/java/com/unascribed/fabrication/mixin/f_balance/pickup_skeleton_arrows/MixinAbstractSkeletonEntity.java]: no, purely to method args. annotating variable/classes is fine. 
iirc arg annotations ware added later and forgeryTools just can't parse them.

unless they changed shadow so you can do `public void create(@Shadow obj)` ? but i'm not aware of that

File: src/main/java/com/unascribed/fabrication/mixin/f_balance/pickup_skeleton_arrows/MixinArrowEntity.java
Comment (SFort) [reply src/main/java/com/unascribed/fabrication/mixin/f_balance/pickup_skeleton_arrows/MixinArrowEntity.java:1]: neat

**Comment by SFort** [comment]
> That did fix the injection failures in `MixinAnvilScreenHandler` for `swap_conflicting_enchants`, but `MixinEntitySummonArgumentType` for `legacy_command_syntax` is still causing injection failures. There aren't even any mixins that target the same class for that one...

well.... yeah.... your target is static, you can't inject a local method into a static function

to be fair FabInjector really should just tell you that instead of the unclear error it throws, that is on me. but that injection cannot work

**Comment by unilock** [comment]
> well.... yeah.... your target is static, you can't inject a local method into a static function

Oops. That should be fixed now.

File: src/main/java/com/unascribed/fabrication/mixin/b_utility/legacy_command_syntax/MixinEntitySummonArgumentType.java [hunk 12]
```diff
@@ -21,19 +22,19 @@
 @Mixin(RegistryEntryArgumentType.class)
 @EligibleIf(configAvailable="*.legacy_command_syntax")
 public class MixinEntitySummonArgumentType {
-	private boolean isNotEntityArgument = true;
+	private static final ThreadLocal<Boolean> fabrication$isNotEntityArgument = ThreadLocal.withInitial(() -> true);
```
Comment (SFort) [inline src/main/java/com/unascribed/fabrication/mixin/b_utility/legacy_command_syntax/MixinEntitySummonArgumentType.java:25]: imo the ThreadLocal usage in `blinking_drops/MixinItemEntityRendererVanilla` is already kinda iffy.
(needs to be tested, but should probably be replaced)

but i'm pretty sure this would cause issues, did you test it?

since this would, set the bool per `Thread` not per `Object` in `<init>` ! 

this injection has gotten a lot harder todo in this version...

File: src/main/java/com/unascribed/fabrication/mixin/i_woina/no_experience/MixinEnchantmentScreen.java [hunk 13]
```diff
@@ -60,17 +59,18 @@ public List<Text> removeLevelText(List<Text> original){
...
-	public boolean fabrication$noXpHijackDrawTexture(DrawContext drawContext, Identifier texture, int x, int y, int u, int v) {
-		if (FabConf.isEnabled("*.no_experience") && (v == 223 || v == 239)) {
-			if (v == 223) {
-				drawContext.drawText(textRenderer, ""+((u/16)+1), x+98, y+8, 0x5577FF, true);
-			}
-			return true;
-		}
-		return false;
-	}
+	// TODO: I have no idea what this is supposed to be doing.
```
[the remark this answers is not in the document]
Comment (SFort) [reply src/main/java/com/unascribed/fabrication/mixin/i_woina/no_experience/MixinEnchantmentScreen.java:62]: turns out this is a pre 1.21 defect

**Comment by SFort** [comment]
I deleted some comments, just because reviews are really annoying to browse.

Yet again i'm out of time, but i fixed some things, removed fapi from `build.gradle` only, so the last commit does not compile. i think i'd be good to get an alpha through soon. i've set some features to "Not Ported" for now ("Old Armor Scale", "Old Armor", "Block Logo", "Legible Signs", "Fullres Banner Shields", "Canhit").

Charm fix is being dropped. i considered doing it accross all MC versions just as a screw you to the Charm devs, but i really don't wan't to fuck over the user unnecessarily.

in a nutshell, todo then merge:
- [x] splitup the networking again to what it used to be (probably done, untested)
- [ ] fix or remove enchantment/datapack related features
- [ ] pull from 1.21 branch (has some fixes)

**Comment by unilock** [comment]
Thank you for taking the reigns while I've been lacking the motivation to do so myself :D

For the sake of simplicity, I think it might be best to just be rid of the troublesome enchantment-related features. (except maybe Bedrock Impaling? I don't think that one can be done with a datapack as easily)

**Comment by SFort** [comment]
i'd just cut all the enchant features for now. it's unfortunate, but if mojang is making things a pain, we can wrangle with it another day, if your up for it. (i don't think there's a point in only porting one or two enchantment features since porting one probably means all the others would work aswell)

soo all that's left is removing enchants, pulling fixes from 3.0/1.21, minimal testing and then we're good to merge. i might have time to do that sunday if you're still outa motivation. but no promises ;p

**Comment by SFort** [comment]
I don't know if i'll have the time today, but the enchantment features aren't being dropped

**Comment by err2ror** [comment]
26.1/26.2....?
